File: csharp/1584-Min-Cost-to-Connect-All-Points.cs
public class Solution
{
    //T: O(N^2 LogN)
    public int MinCostConnectPoints(int[][] points)
    {
        //creating the ajacency list.
        var N = points.Length;
        var dictionary = new Dictionary<int, List<Tuple<int, int>>>(); //Cost and Node

        if (N == 1)
            return 0;

        for (var i = 0; i < N; i++)
        {
            dictionary.Add(i, new List<Tuple<int, int>>());
        }

        for (var i = 0; i < N; i++)
        {
            var x1 = points[i][0];
            var y1 = points[i][1];

            for (var j = i + 1; j < N; j++)
            {
                var x2 = points[j][0];
                var y2 = points[j][1];

                var dist = Math.Abs(x2 - x1) + Math.Abs(y1 - y2);

                dictionary[j].Add(new Tuple<int, int>(dist, i));
                dictionary[i].Add(new Tuple<int, int>(dist, j));

            }
        }

        var res = 0;
        //Prim's
        var visited = new HashSet<int>();
        var minHeap = new PriorityQueue<(int, int), int>(); // Cost and Node
        minHeap.Enqueue((0, 0), 0);

        while (minHeap.Count > 0)
        {
            var (cost, point) = minHeap.Dequeue();
            if (visited.Contains(point))
                continue;
            res += cost;
            visited.Add(point);
            var adj = dictionary[point];

            for (var i = 0; i < adj.Count; i++)
            {
                if (!visited.Contains(adj[i].Item2))
                    minHeap.Enqueue((adj[i].Item1, adj[i].Item2), adj[i].Item1);
            }
        }

        return res;
    }
}

File: javascript/1584-min-cost-to-connect-all-points.js
/**
 * Prim's algorithm
 * https://leetcode.com/problems/min-cost-to-connect-all-points/solution/
 * @param {number[][]} points
 * @return {number}
 */
const minCostConnectPoints = (points) => {
    const isBaseCase = ((points.length === 0) || (1000 <= points.length));
    if (isBaseCase) return 0;

    const { graph, seen, minHeap } = buildGraph(points);

    return search(points, graph, seen, minHeap);
};

const initGraph = (points) => ({
    graph: new Array(points.length).fill().map(() => []),
    seen: new Array(points.length).fill(false),
    minHeap: new MinPriorityQueue()
})

const buildGraph = (points) => {
    const { graph, seen, minHeap } = initGraph(points);

    for (let src = 0; src < (points.length - 1); src++) {
        for (let dst = (src + 1); (dst < points.length); dst++) {
            const cost = getCost(points, src, dst);

            graph[src].push([ dst, cost ]);
            graph[dst].push([ src, cost ]);
        }
    }

    const [ src, cost, priority ] = [ 0, 0, 0 ];
    const node = [ src, cost ];

    minHeap.enqueue(node, priority);

    return { graph, seen, minHeap };
}

const getCost = (points, src, dst) => {
    const [ [ x1, y1 ], [ x2, y2 ] ] = [ points[src], points[dst] ];

    return (Math.abs(x1 - x2) + Math.abs(y1 - y2));
}

const search = (points, graph, seen, minHeap, nodeCount = 0, cost = 0) => {
    while (nodeCount < points.length) {
        let [ src, srcCost ] = minHeap.dequeue().element;

        if (seen[src]) continue;
        seen[src] = true;

        cost += srcCost;
        nodeCount += 1;

        checkNeighbors(graph, src, seen, minHeap);
    }

    return cost;
}

const checkNeighbors = (graph, src, seen, minHeap) => {
    for (const [ dst, dstCost ] of graph[src]) {
        if (seen[dst]) continue;

        minHeap.enqueue([ dst, dstCost ], dstCost);
    }
}